### PR TITLE
Jobs: .With() extension methods.

### PR DESCRIPTION
**Pull request by ig-sinicyn**
As [requested](https://github.com/PerfDotNet/BenchmarkDotNet/pull/286#issuecomment-255080585) by @adamsitnik 

Changelist:
- Jobs: WithXxx() extension methods added back, docs updated.
- Jobs, minor bug fixed: Id was cleared on explicit JobMode assignment + test for it (`Test03IdDoesNotFlow`).
- Jobs.Accuracy, typo: AnaylyzeLaunchVariance => AnalyzeLaunchVariance, docs updated.

### CODE REVIEW

**Comment by AndreyAkinshin** [comment]
In the original design, the `WithXXX()` methods don't change the original job, it always creates a copy of the original job with requested change. Could we keep this approach?

**Comment by ig-sinicyn** [comment]
@AndreyAkinshin  yep, I've think about that, but it feels very counterintuitive.

There're mutable properties, there's .Apply() method that updates the instance of the job and there're .With() methods that creates a new copy of the Job on each call. Kinda Js/PHP style : )

I'd prefer to keep standard Create-Modify-Freeze-Reuse approach until there're strong reasons against it.

**Comment by AndreyAkinshin** [comment]
Ok, let's consider the following situation: I want to have a dry job for Mono. Before the refactoring, I could take the predefined `Dry` job and create a DryMono job in the following way: `Job.Dry.With(Runtime.Mono)`. It was the main use-case of the `WithXXX()` methods.

**Comment by ig-sinicyn** [comment]
@AndreyAkinshin 

Ok, I've fixed the `.With()` method logic. Should not break compatibility now.

**Comment by ig-sinicyn** [comment]
Oops, last minute fix had broken smth.

File: src/BenchmarkDotNet.Core/Jobs/JobExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using BenchmarkDotNet.Toolchains;
+using BenchmarkDotNet.Characteristics;
+using BenchmarkDotNet.Engines;
+using BenchmarkDotNet.Environments;
+using BenchmarkDotNet.Horology;
+
+namespace BenchmarkDotNet.Jobs
+{
+    public static class JobExtensions
```
Comment (ig-sinicyn) [inline src/BenchmarkDotNet.Core/Jobs/JobExtensions.cs]: @AndreyAkinshin 
yep, but each call will incur additional allocation due to closure on WithXxx() arg.
If that's ok - will do.

Comment (AndreyAkinshin) [inline src/BenchmarkDotNet.Core/Jobs/JobExtensions.cs]: It's ok. When we are working with Jobs, we usually don't care about allocations. =)

Comment (ig-sinicyn) [inline src/BenchmarkDotNet.Core/Jobs/JobExtensions.cs]: Ok, fixed 'n' pushed

**Pull request merged**